jax/_src/numpy/lax_numpy.py

```diff
@@ -2479,6 +2479,14 @@ def diagonal(a, offset=0, axis1: int = 0, axis2: int = 1):
 
   diag_size = _max(0, _min(a_shape[axis1] + _min(offset, 0),
                            a_shape[axis2] - _max(offset, 0)))
+  # Skip index normalization for offset == 0
+  if offset == 0 and diag_size > 0:
+    idx_dtype = int64 if diag_size >= (1 << 31) else int_
+    idx = lax.broadcasted_iota(idx_dtype, (diag_size, 2), 0)
+    dims = tuple(range(len(a_shape)))
+    dnums = lax.GatherDimensionNumbers(
```

Comment:
I think I'm going to say "let's leave this the way it was, absent a compelling reason to change it". We can teach XLA how to optimize the HLO better, should it prove worthwhile.

However the fewer direct calls to `lax.gather` in the world the better my life will be... And overall I think I'd rather optimize for simpler *Python* code if the jaxpr isn't that different.

Reply:
I agree, the added complexity is probably not worth it in this case. Currently the only real benefit is that it slightly speeds up compilation time (on CPU by around 20%) and removes the `concatenate` which XLA can't optimise away.